docs/tekton-bundle-contracts.md

```diff
@@ -0,0 +1,97 @@
+<!--
+---
+linkTitle: "Tekton Bundle Contracts"
+weight: 8
+---
+-->
+# Tekton Bundle Contract v0.1
+
+When using a Tekton Bundle in a task or pipeline reference, the OCI artifact backing the
+bundle must adhere to the following contract.
+
+## Contract
+
+Only Tekton CRDs (eg, `Task` or `Pipeline`) may reside in a Tekton Bundle used as a Tekton
+bundle reference.
+
+Each layer of the image must map 1:1 with a single Tekton resource.
+
+Each layer must contain all of the following annotations:
+
+- `org.opencontainers.image.title` => `ObjectMeta.Name` of the resource
+- `dev.tekton.image.kind` => `TypeMeta.Kind` of the resource, all lowercased (eg, `task`)
+- `dev.tekton.image.apiVersion` => `TypeMeta.APIVersion` of the resource (eg 
+"tekton.dev/v1alpha1")  
+
+The union of the { `dev.tekton.image.apiVersion`, `dev.tekton.image.kind`, `org.opencontainers.image.title` }
+annotations on a given layer must be unique among all layers of that image. In practical terms, this means no two
+"tasks" can have the same name for example.
+
+Each layer must map 1:1 with a single Tekton resource (eg Task), be compressed and stored with a supported OCI MIME type 
+*except* for `+zstd` types. For list of the supported types see
+[the official spec](https://github.com/opencontainers/image-spec/blob/master/layer.md#zstd-media-types).
```

Comment:
I am worried that you're using these mime types incorrectly. 

The layer mimes used for images specifically represent [filesystem changesets](https://github.com/opencontainers/image-spec/blob/master/layer.md#image-layer-filesystem-changeset). It looks like you're expecting there to be exactly one resource (file?) per layer. What happens if two layers have the same resource? The mime type you're proposing would suggest that we follow the rules of [applying changesets](https://github.com/opencontainers/image-spec/blob/master/layer.md#applying-changesets), but this suggests otherwise:

> In practical terms, this means no two "tasks" can have the same name for example.

These layer mimetypes also require the contents to be tarred, which isn't called out here.

I would really suggest using a media type that helps you when decoding the content of the layer, e.g. if you can express that it's YAML or JSON, that would be helpful.

I'm also not sure if it's strictly necessary that the layers are compressed? I imagine it's just text, but compression doesn't need to be called out explicitly, as [the suffixes](https://www.iana.org/assignments/media-type-structured-suffix/media-type-structured-suffix.xhtml) are well-defined.

Reply:
The compressed/not compressed is honestly not something I have thought about deeply since as far as I can tell, the library handles the gzip/untar for the user. That is about where my knowledge in all this ends to be honest; I am sure you know way more than I do about this.

All I can say is that my intent is to make this as simple as possible for the Tekton control to parse. The "root" object in a sense that I found capable of attaching metadata to was a "layer" and that figured to be as good as anything else for using to represent a single resource. If they overlap, "theoretically" the contract enforces that this would be invalid because each object can only appear once though obviously we have no way to stop that. The implementation resolves this by ignoring repeats.

So I guess my TL;DR is that I don't actual mind what the type of the layer is because I am just going to read it as raw bytes and try to parse that as JSON or YAML anyways and fail if it can't. So whatever you suggest that moves me in that direction I am all for.

Reply:
With the caveat that I'm not sure if all registries would work with it (generally, they should), I would propose using `application/yaml` or `application/json` as appropriate. If you want to gzip them, add the `+gzip` suffix.

I'm not completely sure how go-containerregistry handles this. I would guess that by default it assumes layers are gzipped, but I'm happy to fix that.

Reply:
+1 to Jon's point. I think the best approach would be to set the layer as application/yaml/json.

Reply:
So I just tried this out and this worked for Docker but did not work for Artifactory (other registry). I think part of the issue is the extent to which the registry supports the full OCI spec. I worry that Artifactory is not alone here in not being fully compliant.

It seems like the safest approach here is to use one of the well-supported default types and the long-term ideal is to have a custom mime type for a Tekton resource the same way that Helm does. Therefore, I am inclined to keep this as is for now as I have tested that this works with multiple registries as of right now.

Reply:
Oh nice find!

Using the existing media types works for me, but if we do that I think we should probably actually use those media types correctly and wrap up these bytes into a tar file. It's definitely a bit more code, but IMO is worth it so any tooling that inspects these images can correctly untar these files.

Reply:
Fair enough. My next PR is to add the tooling to create these images and I believe that is where we will need to add this "tar'ing" logic.

Reply:
> I believe that is where we will need to add this "tar'ing" logic.

Looks like this is how they're read? Need to do the untarring here, as well:
https://github.com/tektoncd/pipeline/blob/81802739f5356f927c4f976a7a3f000fee95b773/pkg/remote/oci/resolver.go#L98-L113

Reply:
It only unzips, it does not tar/untar. We have to do that part ourselves.